internal/configv2/configv2.go

```diff
@@ -0,0 +1,409 @@
+// limitations under the License.
+
+// Package configv2 implements loading of configuration V2 from Viper configuration.
+// The implementation relies on registered factories that allow creating
+// default configuration for each type of receiver/exporter/option.
+package configv2
+
+import (
+	"fmt"
+
+	"github.com/census-instrumentation/opencensus-service/internal/models"
+	"github.com/spf13/viper"
+)
+
+// These are errors that can be returned by Load(). Note that error codes are not part
+// of Load()'s public API, they are for internal unit testing only.
+type configErrorCode int
+
+const (
+	_ configErrorCode = iota // skip 0, start errors codes from 1
+	errUnknownReceiverType
+	errUnknownExporterType
+	errUnknownOptionType
+	errMissingReceiverType
+	errMissingExporterType
+	errMissingOptionType
+	errDuplicateReceiverName
+	errDuplicateExporterName
+	errDuplicateOptionName
+	errMissingPipelines
+	errMissingPipelineName
+	errPipelineMustHaveReceiver
+	errPipelineMustHaveExporter
+	errPipelineMustHaveOperations
+	errPipelineReceiverNotExists
+	errPipelineOptionNotExists
+	errPipelineExporterNotExists
+	errMetricPipelineCannotHaveOperations
+	errUnmarshalError
+)
+
+type configError struct {
+	msg  string          // human readable error message
+	code configErrorCode // internal error code
+}
+
+// ensure configError implements error interface.
+var _ error = (*configError)(nil)
+
+func (e *configError) Error() string {
+	return e.msg
+}
+
+// Load loads a ConfigV2 from Viper
+func Load(v *viper.Viper) (*models.ConfigV2, error) {
+
+	// Unmarshal everything that can be automatically unmarshaled (currently only pipelines)
+	var config models.ConfigV2
+	if err := v.Unmarshal(&config); err != nil {
+		return nil, err
+	}
+
+	// Load the rest of entities that are not automatically unmarshaled: receivers, exporters, options.
+
+	receivers, err := loadReceivers(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Receivers = receivers
+
+	exporters, err := loadExporters(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Exporters = exporters
+
+	options, err := loadOptions(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Options = options
+
+	// Config is loaded. Now validate it.
+
+	if err := validateConfig(&config); err != nil {
+		return nil, err
+	}
+
+	return &config, nil
+}
+
+func loadReceivers(v *viper.Viper) ([]models.ReceiverCfg, error) {
+	// Get the list of all "receivers" items from config source
+	subs, err := getViperSubSequenceOfMaps(v, models.ReceiversKeyName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate over receivers and create a config for each
+	receivers := make([]models.ReceiverCfg, 0)
+	for _, sub := range subs {
+		// Find receiver factory based on "type" that we read from config source
+		typeStr := sub.GetString(models.TypeKeyName)
+		if typeStr == "" {
+			return nil, &configError{code: errMissingReceiverType, msg: fmt.Sprintf("missing receiver type")}
```

Comment:
@tigrannajaryan nit: just noticed this and others below: unneeded `fmt.Sprintf` since string is a constant.

Reply:
Fixed.